pkg/controllers/cronfederatedhpa/cronfederatedhpa_controller.go

```diff
@@ -0,0 +1,207 @@
+			return controllerruntime.Result{}, err
+		}
+		newRuleSets.Insert(rule.Name)
+	}
+
+	// If rule is deleted, remove the rule executor from the handler
+	for name := range origRuleSets {
+		if !newRuleSets.Has(name) {
+			c.CronHandler.StopRuleExecutor(req.NamespacedName.String(), name)
+			if err = c.removeCronFHPAHistory(cronFHPA, name); err != nil {
+				return controllerruntime.Result{}, err
+			}
+		}
+	}
+
+	return controllerruntime.Result{}, nil
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CronFHPAController) SetupWithManager(mgr controllerruntime.Manager) error {
+	c.CronHandler = NewCronHandler(mgr.GetClient(), mgr.GetEventRecorderFor(ControllerName))
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&autoscalingv1alpha1.CronFederatedHPA{}).
+		WithOptions(controller.Options{RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RateLimiterOptions)}).
+		Complete(c)
+}
+
+// processCronRule processes the cron rule
+func (c *CronFHPAController) processCronRule(cronFHPA *autoscalingv1alpha1.CronFederatedHPA, rule autoscalingv1alpha1.CronFederatedHPARule) error {
+	cronFHPAKey := helper.GetCronFederatedHPAKey(cronFHPA)
+	ruleOld, exists := c.CronHandler.RuleCronExecutorExists(cronFHPAKey, rule.Name)
+	if exists {
+		if !equality.Semantic.DeepEqual(ruleOld, rule) {
+			c.CronHandler.StopRuleExecutor(cronFHPAKey, rule.Name)
+		} else {
+			// rule is not changed, ignore it
+			return nil
+		}
+	}
+
+	suspended := helper.IsCronFederatedHPARuleSuspend(rule)
+	if !suspended {
+		if err := c.CronHandler.CreateCronJobForExecutor(cronFHPA, rule); err != nil {
+			klog.Error("Failed to start cron job for executor:", err)
+			return err
+		}
+	}
+
+	return c.updateRuleNextExecutionTime(cronFHPA, rule, suspended)
+}
+
+// updateRuleNextExecutionTime updates the next execution time of the rule
+func (c *CronFHPAController) updateRuleNextExecutionTime(cronFHPA *autoscalingv1alpha1.CronFederatedHPA,
+	rule autoscalingv1alpha1.CronFederatedHPARule, suspend bool) error {
+	ctx := context.Background()
+	exists := false
+
+	var nextExecutionTime *metav1.Time
+	if !suspend {
+		// If rule is not suspended, we should set the nextExecutionTime filed, or the nextExecutionTime will be nil
+		next, err := helper.GetCronFederatedHPARuleNextExecutionTime(rule)
+		if err != nil {
+			return err
+		}
+		nextExecutionTime = &metav1.Time{Time: next}
+	}
+
+	for index, history := range cronFHPA.Status.ExecutionHistories {
+		if history.RuleName == rule.Name {
+			exists = true
+			cronFHPA.Status.ExecutionHistories[index].NextExecutionTime = nextExecutionTime
+			break
+		}
+	}
+
+	if !exists {
+		ruleHistory := autoscalingv1alpha1.ExecutionHistory{
+			RuleName:          rule.Name,
+			NextExecutionTime: nextExecutionTime,
+		}
+		cronFHPA.Status.ExecutionHistories = append(cronFHPA.Status.ExecutionHistories, ruleHistory)
+	}
+
+	if err := c.Client.Status().Update(ctx, cronFHPA); err != nil {
```

Comment:
`cronFHPA` may remain unchanged. Only when `cronFHPA` is modified, should `Update()` be called.

Reply:
This function will be called only when rules changed(updated or added), So I think it meets `Only when cronFHPA is modified, should Update() be called`.

Reply:
> This function will be called only when rules changed(updated or added), 

Yes, if only `TargetMinReplicas` is modified and `Schedule` remains unchanged, there is no need to recalculate `NextExecutionTime`.

Reply:
I think we need, becuase the data changed, we need to stop the existing cron and start one new the the new `TargetMinReplicas `. Once the corn is restart again, we should update the NextExecutionTime